include/aws/sdkutils/endpoints/private/ruleset_types_impl.h

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_SDKUTILS_ENDPOINTS_RULESET_TYPES_IMPL_H
+#define AWS_SDKUTILS_ENDPOINTS_RULESET_TYPES_IMPL_H
+
+#include <aws/common/ref_count.h>
+#include <aws/sdkutils/endpoints/ruleset.h>
+
+enum aws_endpoints_rule_type { AWS_ENDPOINTS_RULE_ENDPOINT, AWS_ENDPOINTS_RULE_ERROR, AWS_ENDPOINTS_RULE_TREE };
+
+enum aws_endpoints_url_type { AWS_ENDPOINTS_URL_TEMPLATE, AWS_ENDPOINTS_URL_REFERENCE, AWS_ENDPOINTS_URL_FUNCTION };
+
+enum aws_endpoints_error_type {
+    AWS_ENDPOINTS_ERROR_TEMPLATE,
+    AWS_ENDPOINTS_ERROR_REFERENCE,
+    AWS_ENDPOINTS_ERROR_FUNCTION
+};
+
+enum aws_endpoints_expr_type {
+    AWS_ENDPOINTS_EXPR_STRING,
+    AWS_ENDPOINTS_EXPR_NUMBER,
+    AWS_ENDPOINTS_EXPR_BOOLEAN,
+    AWS_ENDPOINTS_EXPR_ARRAY,
+    AWS_ENDPOINTS_EXPR_REFERENCE,
+    AWS_ENDPOINTS_EXPR_FUNCTION
+};
+
+struct aws_endpoints_parameter {
+    struct aws_allocator *allocator;
+
+    enum aws_endpoints_parameter_value_type type;
+    struct aws_string *built_in;
+
+    union {
+        struct aws_string *string;
+        bool boolean;
+    } default_value;
+
+    bool is_required;
+    struct aws_string *documentation;
+    bool is_deprecated;
+    struct aws_string *deprecated_message;
+    struct aws_string *deprecated_since;
+};
+
+struct aws_endpoints_ruleset {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+
+    struct aws_array_list rules;
+
+    struct aws_string *version;
+    struct aws_string *service_id;
+    struct aws_hash_table *parameters;
+};
+
+struct aws_endpoints_reference {
+    struct aws_allocator *allocator;
+    struct aws_string *ref;
+};
+
+struct aws_endpoints_fn;
+
+struct aws_endpoints_expr {
+    struct aws_allocator *allocator;
+    enum aws_endpoints_expr_type type;
+    union {
+        struct aws_string *string;
+        int32_t number;
+        bool boolean;
+        struct aws_array_list array;
+        struct aws_endpoints_reference *reference;
+        struct aws_endpoints_function *function;
+    } e;
+};
+
+struct aws_endpoints_function {
+    struct aws_allocator *allocator;
+    struct aws_string *fn; /* TODO: precompute hash or enum value to avoid lookup on every eval */
```

Comment:
pretty much the same comments as the one about `struct aws_endpoints_reference`

we could just store this by-value within the union, instead of doing another heap allocation, and we probably don't need to store the allocator since we're only destroying this while destroying the `aws_endpoints_expr` that owns it